File: AI-CodeEditor/securefix-ide/backend/app/services/fix_generator_service.py
"""
Fix Generator Service - Core AI Fix Engine

Handles streaming AI fix generation using Azure OpenAI or standard OpenAI.
Integrates with LangChain for structured prompting and token streaming.
"""

import re
import time
import logging
from typing import Dict, Any, Optional, List, AsyncGenerator, Tuple
from datetime import datetime

from langchain_openai import AzureChatOpenAI, ChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage
from langchain_core.prompts import ChatPromptTemplate

from config.settings import settings
from .prompt_templates import (
    SYSTEM_PROMPT_FIX_GENERATION,
    SYSTEM_PROMPT_CHAT,
    build_fix_prompt,
    build_chat_context_prompt,
    get_quick_action_prompt
)
from .context_extraction_service import ContextExtractionService, CodeContext

logger = logging.getLogger(__name__)


class FixGeneratorService:
    """
    AI-powered security fix generator with streaming support.

    Features:
    - Streaming token-by-token generation (Cursor-style)
    - Security-focused prompts with CWE/OWASP context
    - Search/replace block parsing for precise code application
    - Support for both Azure OpenAI and standard OpenAI
    """

    def __init__(self):
        """Initialize the fix generator with configured LLM."""
        self.context_extractor = ContextExtractionService(
            context_lines_before=settings.CONTEXT_LINES_BEFORE,
            context_lines_after=settings.CONTEXT_LINES_AFTER
        )
        self.llm = self._create_llm()

    def _create_llm(self):
        """Create the appropriate LLM client based on configuration."""
        if settings.AZURE_OPENAI_API_KEY and settings.AZURE_OPENAI_ENDPOINT:
            logger.info("Using Azure OpenAI for fix generation")
            return AzureChatOpenAI(
                azure_endpoint=settings.AZURE_OPENAI_ENDPOINT,
                api_key=settings.AZURE_OPENAI_API_KEY,
                api_version=settings.AZURE_OPENAI_API_VERSION,
                deployment_name=settings.AZURE_OPENAI_DEPLOYMENT,
                temperature=settings.LLM_TEMPERATURE,
                max_tokens=settings.LLM_MAX_TOKENS,
                streaming=settings.LLM_STREAMING
            )
        elif settings.OPENAI_API_KEY:
            logger.info("Using OpenAI for fix generation")
            return ChatOpenAI(
                api_key=settings.OPENAI_API_KEY,
                model=settings.OPENAI_MODEL,
                temperature=settings.LLM_TEMPERATURE,
                max_tokens=settings.LLM_MAX_TOKENS,
                streaming=settings.LLM_STREAMING
            )
        else:
            logger.warning("No LLM API key configured - fix generation will fail")
            return None

    async def generate_fix_streaming(
        self,
        vulnerability: Dict[str, Any],
        code_context: Dict[str, Any],
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """
        Generate a security fix with streaming response.

        Args:
            vulnerability: Vulnerability information dict with cwe, severity, description
            code_context: Code context dict with file_path, start_line, end_line,
                         vulnerable_code, surrounding_code, imports, framework

        Yields:
            Dict with type='fix_chunk', content=str, done=bool
            On completion: includes search_blocks and replace_blocks
        """
        if not self.llm:
            yield {
                "type": "fix_error",
                "message": "LLM not configured. Please set AZURE_OPENAI_API_KEY or OPENAI_API_KEY.",
                "vulnerability_id": vulnerability.get("id", "unknown")
            }
            return

        start_time = time.time()
        full_content = ""
        prompt_tokens = 0
        completion_tokens = 0

        try:
            # Build the fix generation prompt
            language = code_context.get("language", "python")
            framework = code_context.get("framework")

            user_prompt = build_fix_prompt(
                vulnerability=vulnerability,
                code_context=code_context,
                framework=framework,
                language=language
            )

            messages = [
                SystemMessage(content=SYSTEM_PROMPT_FIX_GENERATION),
                HumanMessage(content=user_prompt)
            ]

            logger.info(f"Starting fix generation for vulnerability: {vulnerability.get('cwe', 'unknown')}")

            # Stream the response
            async for chunk in self.llm.astream(messages):
                if hasattr(chunk, "content") and chunk.content:
                    full_content += chunk.content
                    yield {
                        "type": "fix_chunk",
                        "content": chunk.content,
                        "done": False,
                        "vulnerability_id": vulnerability.get("id", "unknown")
                    }

                # Track token usage if available
                if hasattr(chunk, "usage_metadata") and chunk.usage_metadata:
                    prompt_tokens = chunk.usage_metadata.get("input_tokens", 0)
                    completion_tokens = chunk.usage_metadata.get("output_tokens", 0)

            # Parse search/replace blocks from the complete response
            search_blocks, replace_blocks = self._parse_search_replace_blocks(full_content)

            generation_time_ms = int((time.time() - start_time) * 1000)

            logger.info(
                f"Fix generation completed: {len(search_blocks)} blocks, "
                f"{generation_time_ms}ms, {prompt_tokens}+{completion_tokens} tokens"
            )

            # Send completion message
            yield {
                "type": "fix_chunk",
                "content": "",
                "done": True,
                "vulnerability_id": vulnerability.get("id", "unknown"),
                "full_content": full_content,
                "search_blocks": search_blocks,
                "replace_blocks": replace_blocks,
                "generation_time_ms": generation_time_ms,
                "prompt_tokens": prompt_tokens,
                "completion_tokens": completion_tokens
            }

        except Exception as e:
            logger.error(f"Error during fix generation: {e}", exc_info=True)
            yield {
                "type": "fix_error",
                "message": str(e),
                "vulnerability_id": vulnerability.get("id", "unknown")
            }

    async def generate_fix(
        self,
        vulnerability: Dict[str, Any],
        code_context: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Generate a security fix (non-streaming, returns complete result).

        Args:
            vulnerability: Vulnerability information dict
            code_context: Code context dict

        Returns:
            Dict with raw_content, search_blocks, replace_blocks, metadata
        """
        full_content = ""
        result = {}

        async for chunk in self.generate_fix_streaming(vulnerability, code_context):
            if chunk["type"] == "fix_chunk":
                if chunk.get("done"):
                    result = chunk
                else:
                    full_content += chunk.get("content", "")
            elif chunk["type"] == "fix_error":
                raise Exception(chunk["message"])

        return result

    def _parse_search_replace_blocks(self, content: str) -> Tuple[List[str], List[str]]:
        """
        Parse search/replace blocks from LLM output.

        Expected format:
        <<<SEARCH
        original code here
        >>>
        <<<REPLACE
        fixed code here
        >>>

        Returns:
            Tuple of (search_blocks, replace_blocks)
        """
        search_blocks = []
        replace_blocks = []

        # Pattern to match SEARCH/REPLACE blocks
        pattern = r"<<<SEARCH\n([\s\S]*?)\n>>>\n<<<REPLACE\n([\s\S]*?)\n>>>"

        matches = re.findall(pattern, content)
        for search, replace in matches:
            search_blocks.append(search.strip())
            replace_blocks.append(replace.strip())

        # Also try alternative formats
        if not search_blocks:
            # Try ```search and ```replace format
            alt_pattern = r"```search\n([\s\S]*?)\n```\n```replace\n([\s\S]*?)\n```"
            matches = re.findall(alt_pattern, content, re.IGNORECASE)
            for search, replace in matches:
                search_blocks.append(search.strip())
                replace_blocks.append(replace.strip())

        return search_blocks, replace_blocks

    def _check_for_context_request(self, content: str) -> Optional[List[Dict[str, Any]]]:
        """
        Check if the AI is requesting more context.

        The AI may respond with:
        NEED_CONTEXT: function:validate_user, class:UserService, imports, lines:50-70

        Returns:
            List of context requests or None if no request
        """
        # Check for context request pattern
        pattern = r"NEED_CONTEXT:\s*(.+)"
        match = re.search(pattern, content, re.IGNORECASE)

        if not match:
            return None

        requests = []
        request_str = match.group(1).strip()

        for item in request_str.split(","):
            item = item.strip()
            if item.startswith("function:"):
                requests.append({"type": "function", "name": item[9:].strip()})
            elif item.startswith("class:"):
                requests.append({"type": "class", "name": item[6:].strip()})
            elif item == "imports":
                requests.append({"type": "imports"})
            elif item.startswith("lines:"):
                range_str = item[6:].strip()
                if "-" in range_str:
                    start, end = range_str.split("-")
                    requests.append({"type": "lines", "start": int(start), "end": int(end)})
            elif item == "related":
                requests.append({"type": "related_functions"})

        return requests if requests else None

    async def chat_streaming(
        self,
        message: str,
        context: Optional[Dict[str, Any]] = None,
        history: Optional[List[Dict[str, str]]] = None
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """
        Generate a chat response with streaming.

        Args:
            message: User's message
            context: Optional context (vulnerability, current_file, recent_fix)
            history: Optional conversation history

        Yields:
            Dict with type='chat_chunk', content=str, done=bool
        """
        if not self.llm:
            yield {
                "type": "chat_error",
                "message": "LLM not configured."
            }
            return

        try:
            # Build context-aware prompt
            context = context or {}
            user_prompt = build_chat_context_prompt(
                vulnerability=context.get("vulnerability"),
                current_file=context.get("current_file"),
                recent_fix=context.get("recent_fix"),
                user_message=message
            )

            messages = [SystemMessage(content=SYSTEM_PROMPT_CHAT)]

            # Add conversation history
            if history:
                for msg in history[-10:]:  # Keep last 10 messages
                    if msg["role"] == "user":
                        messages.append(HumanMessage(content=msg["content"]))
                    elif msg["role"] == "assistant":
                        messages.append(SystemMessage(content=msg["content"]))

            messages.append(HumanMessage(content=user_prompt))

            full_content = ""
            async for chunk in self.llm.astream(messages):
                if hasattr(chunk, "content") and chunk.content:
                    full_content += chunk.content
                    yield {
                        "type": "chat_chunk",
                        "content": chunk.content,
                        "done": False
                    }

            yield {
                "type": "chat_chunk",
                "content": "",
                "done": True,
                "full_content": full_content
            }

        except Exception as e:
            logger.error(f"Error during chat: {e}", exc_info=True)
            yield {
                "type": "chat_error",
                "message": str(e)
            }

    async def quick_action(
        self,
        action: str,
        vulnerability: Dict[str, Any]
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """
        Execute a quick action (explain, owasp, alternative, test_cases, impact).

        Args:
            action: Action type
            vulnerability: Vulnerability information

        Yields:
            Dict with type='chat_chunk', content=str, done=bool
        """
        if not self.llm:
            yield {
                "type": "chat_error",
                "message": "LLM not configured."
            }
            return

        try:
            prompt = get_quick_action_prompt(action, vulnerability)
            messages = [
                SystemMessage(content=SYSTEM_PROMPT_CHAT),
                HumanMessage(content=prompt)
            ]

            full_content = ""
            async for chunk in self.llm.astream(messages):
                if hasattr(chunk, "content") and chunk.content:
                    full_content += chunk.content
                    yield {
                        "type": "chat_chunk",
                        "content": chunk.content,
                        "done": False
                    }

            yield {
                "type": "chat_chunk",
                "content": "",
                "done": True,
                "full_content": full_content,
                "action": action
            }

        except Exception as e:
            logger.error(f"Error during quick action '{action}': {e}", exc_info=True)
            yield {
                "type": "chat_error",
                "message": str(e)
            }

    def extract_code_context(
        self,
        file_path: str,
        start_line: int,
        end_line: int,
        repo_path: Optional[str] = None
    ) -> CodeContext:
        """
        Extract code context for fix generation.

        Args:
            file_path: Path to the file
            start_line: Starting line number
            end_line: Ending line number
            repo_path: Optional repository root path

        Returns:
            CodeContext object
        """
        return self.context_extractor.extract_context(
            file_path=file_path,
            start_line=start_line,
            end_line=end_line,
            repo_path=repo_path
        )

    def extract_code_context_from_content(
        self,
        content: str,
        file_path: str,
        start_line: int,
        end_line: int
    ) -> CodeContext:
        """
        Extract code context from provided content.

        Args:
            content: Full file content
            file_path: File path for language detection
            start_line: Starting line number
            end_line: Ending line number

        Returns:
            CodeContext object
        """
        return self.context_extractor.extract_context_from_content(
            content=content,
            file_path=file_path,
            start_line=start_line,
            end_line=end_line
        )


# Singleton instance
fix_generator = FixGeneratorService()

File: AI-CodeEditor/securefix-ide/core/src/services/FixGeneratorService.ts
/**
 * FixGeneratorService
 *
 * Constructs security-aware prompts with code context,
 * manages streaming from backend, and parses fix output.
 *
 * Connects to the SecureFix AI Fix Engine backend (Phase 2.3)
 * Default: http://localhost:8000 (FastAPI backend)
 */

import type { Vulnerability } from '../types/vulnerability.js';
import type { CodeContext, FrameworkInfo } from '../types/code-context.js';
import type { SearchReplaceBlock } from '../types/diff-zone.js';

export interface FixGeneratorConfig {
  /** Backend URL (default: http://localhost:8000) */
  backendUrl: string;
  /** Optional API key for authentication */
  apiKey?: string;
  /** WebSocket URL override (defaults to backendUrl with ws:// protocol) */
  wsUrl?: string;
}

export interface FixStreamCallbacks {
  onChunk: (content: string, done: boolean) => void;
  onComplete: (fullContent: string, blocks: SearchReplaceBlock[]) => void;
  onError: (error: Error) => void;
}

export interface FixGenerationMetadata {
  generationTimeMs?: number;
  promptTokens?: number;
  completionTokens?: number;
}

export class FixGeneratorService {
  private config: FixGeneratorConfig;
  private activeConnections: Map<string, WebSocket> = new Map();

  constructor(config: FixGeneratorConfig) {
    // Default to the AI Fix Engine backend
    this.config = {
      backendUrl: config.backendUrl || 'http://localhost:8000',
      apiKey: config.apiKey,
      wsUrl: config.wsUrl
    };
  }

  /**
   * Generate a fix using streaming WebSocket connection
   * Connects to: /api/fix/ws/fix
   */
  async generateFix(
    vulnerability: Vulnerability,
    codeContext: CodeContext,
    callbacks: FixStreamCallbacks
  ): Promise<void> {
    const wsUrl = this.config.wsUrl || this.config.backendUrl.replace('http', 'ws');
    // Connect to the AI Fix Engine WebSocket endpoint
    const ws = new WebSocket(`${wsUrl}/api/fix/ws/fix`);
    const connectionId = vulnerability.id;

    let fullContent = '';

    ws.onopen = () => {
      // Send fix request with context (matching backend FixRequest schema)
      const request = {
        type: 'fix_request',
        vulnerability: {
          id: vulnerability.id,
          cwe: typeof vulnerability.cwe === 'object' ? vulnerability.cwe.id : vulnerability.cwe,
          severity: vulnerability.severity,
          description: vulnerability.description,
          recommendation: vulnerability.recommendation,
          owasp: vulnerability.owasp?.category
        },
        codeContext: {
          file_path: codeContext.fileUri,
          start_line: codeContext.startLine,
          end_line: codeContext.endLine,
          vulnerable_code: codeContext.vulnerableCode,
          surrounding_code: codeContext.surroundingCode,
          imports: codeContext.imports,
          language: codeContext.language,
          framework: codeContext.framework?.name
        }
      };

      ws.send(JSON.stringify(request));
    };

    ws.onmessage = (event) => {
      const data = JSON.parse(event.data);

      if (data.type === 'fix_chunk') {
        fullContent += data.content || '';
        callbacks.onChunk(data.content || '', data.done);

        if (data.done) {
          // Backend provides parsed blocks, but also parse locally as fallback
          const blocks = data.search_blocks && data.replace_blocks
            ? data.search_blocks.map((search: string, i: number) => ({
                search,
                replace: data.replace_blocks[i] || ''
              }))
            : this.parseSearchReplaceBlocks(fullContent);

          callbacks.onComplete(data.full_content || fullContent, blocks);
          ws.close();
          this.activeConnections.delete(connectionId);
        }
      } else if (data.type === 'fix_error') {
        callbacks.onError(new Error(data.message));
        ws.close();
        this.activeConnections.delete(connectionId);
      }
    };

    ws.onerror = (_event) => {
      callbacks.onError(new Error('WebSocket connection error'));
      this.activeConnections.delete(connectionId);
    };

    ws.onclose = () => {
      this.activeConnections.delete(connectionId);
    };

    this.activeConnections.set(connectionId, ws);
  }

  /**
   * Cancel an ongoing fix generation
   */
  cancelFix(vulnerabilityId: string): void {
    const ws = this.activeConnections.get(vulnerabilityId);
    if (ws) {
      ws.close();
      this.activeConnections.delete(vulnerabilityId);
    }
  }

  /**
   * Parse search/replace blocks from LLM output
   *
   * Expected format:
   * <<<SEARCH
   * original code here
   * >>>
   * <<<REPLACE
   * fixed code here
   * >>>
   */
  parseSearchReplaceBlocks(content: string): SearchReplaceBlock[] {
    const blocks: SearchReplaceBlock[] = [];
    const regex = /<<<SEARCH\n([\s\S]*?)\n>>>\n<<<REPLACE\n([\s\S]*?)\n>>>/g;

    let match;
    while ((match = regex.exec(content)) !== null) {
      blocks.push({
        search: match[1].replace(/\r\n/g, '\n').replace(/^\n+|\n+$/g, ''),
        replace: match[2].replace(/\r\n/g, '\n').replace(/^\n+|\n+$/g, '')
      });
    }

    return blocks;
  }

  /**
   * Build the prompt for fix generation
   */
  buildPrompt(
    vulnerability: Vulnerability,
    codeContext: CodeContext
  ): string {
    const frameworkHints = codeContext.framework
      ? this.getFrameworkHints(codeContext.framework)
      : '';

    return `You are a security expert. Fix the following vulnerability with minimal code changes.
Preserve existing logic and coding style.

## Vulnerability Information
- CWE: ${vulnerability.cwe.id} - ${vulnerability.cwe.name}
- Severity: ${vulnerability.severity.toUpperCase()}
- Description: ${vulnerability.description}
${vulnerability.owasp ? `- OWASP Category: ${vulnerability.owasp.category} - ${vulnerability.owasp.name}` : ''}

## Recommendation
${vulnerability.recommendation}

## Code Context
Language: ${codeContext.language}
File: ${codeContext.fileUri}
Vulnerable lines: ${codeContext.startLine}-${codeContext.endLine}
${frameworkHints}

### Imports
\`\`\`
${codeContext.imports.join('\n')}
\`\`\`

### Surrounding Code
\`\`\`${codeContext.language}
${codeContext.surroundingCode}
\`\`\`

### Vulnerable Code
\`\`\`${codeContext.language}
${codeContext.vulnerableCode}
\`\`\`

## Output Format
Provide the fix using search/replace blocks for precise application:

<<<SEARCH
exact original code to find
>>>
<<<REPLACE
fixed secure code
>>>

Important:
- Make minimal changes to fix the vulnerability
- Maintain existing code style and indentation
- Only include the exact code that needs to change in SEARCH blocks
- Do not add unnecessary comments or refactoring`;
  }

  /**
   * Get framework-specific security hints
   */
  private getFrameworkHints(framework: FrameworkInfo): string {
    return `
### Framework: ${framework.name}
Security Patterns:
${framework.securityPatterns.map(p => `- ${p}`).join('\n')}`;
  }

  /**
   * Chat with AI about a vulnerability
   * Connects to: /api/fix/ws/chat
   */
  async chat(
    message: string,
    context: {
      vulnerability?: Vulnerability;
      currentFile?: string;
      recentFix?: string;
    },
    callbacks: {
      onChunk: (content: string, done: boolean) => void;
      onComplete: (fullContent: string) => void;
      onError: (error: Error) => void;
    }
  ): Promise<void> {
    const wsUrl = this.config.wsUrl || this.config.backendUrl.replace('http', 'ws');
    const ws = new WebSocket(`${wsUrl}/api/fix/ws/chat`);
    const connectionId = `chat_${Date.now()}`;

    let fullContent = '';

    ws.onopen = () => {
      const request = {
        type: 'chat_message',
        content: message,
        context: {
          vulnerability: context.vulnerability ? {
            id: context.vulnerability.id,
            cwe: typeof context.vulnerability.cwe === 'object'
              ? context.vulnerability.cwe.id
              : context.vulnerability.cwe,
            severity: context.vulnerability.severity,
            description: context.vulnerability.description,
            file_path: context.vulnerability.filePath
          } : undefined,
          current_file: context.currentFile,
          recent_fix: context.recentFix
        }
      };

      ws.send(JSON.stringify(request));
    };

    ws.onmessage = (event) => {
      const data = JSON.parse(event.data);

      if (data.type === 'chat_chunk') {
        fullContent += data.content || '';
        callbacks.onChunk(data.content || '', data.done);

        if (data.done) {
          callbacks.onComplete(data.full_content || fullContent);
          ws.close();
          this.activeConnections.delete(connectionId);
        }
      } else if (data.type === 'chat_error') {
        callbacks.onError(new Error(data.message));
        ws.close();
        this.activeConnections.delete(connectionId);
      }
    };

    ws.onerror = () => {
      callbacks.onError(new Error('Chat WebSocket connection error'));
      this.activeConnections.delete(connectionId);
    };

    ws.onclose = () => {
      this.activeConnections.delete(connectionId);
    };

    this.activeConnections.set(connectionId, ws);
  }

  /**
   * Execute a quick action (explain, owasp, alternative, test_cases, impact)
   * Connects to: /api/fix/ws/chat
   */
  async quickAction(
    action: 'explain' | 'owasp' | 'alternative' | 'test_cases' | 'impact',
    vulnerability: Vulnerability,
    callbacks: {
      onChunk: (content: string, done: boolean) => void;
      onComplete: (fullContent: string) => void;
      onError: (error: Error) => void;
    }
  ): Promise<void> {
    const wsUrl = this.config.wsUrl || this.config.backendUrl.replace('http', 'ws');
    const ws = new WebSocket(`${wsUrl}/api/fix/ws/chat`);
    const connectionId = `action_${action}_${Date.now()}`;

    let fullContent = '';

    ws.onopen = () => {
      const request = {
        type: 'quick_action',
        action,
        vulnerability: {
          id: vulnerability.id,
          cwe: typeof vulnerability.cwe === 'object' ? vulnerability.cwe.id : vulnerability.cwe,
          severity: vulnerability.severity,
          description: vulnerability.description,
          file_path: vulnerability.filePath
        }
      };

      ws.send(JSON.stringify(request));
    };

    ws.onmessage = (event) => {
      const data = JSON.parse(event.data);

      if (data.type === 'chat_chunk') {
        fullContent += data.content || '';
        callbacks.onChunk(data.content || '', data.done);

        if (data.done) {
          callbacks.onComplete(data.full_content || fullContent);
          ws.close();
          this.activeConnections.delete(connectionId);
        }
      } else if (data.type === 'chat_error') {
        callbacks.onError(new Error(data.message));
        ws.close();
        this.activeConnections.delete(connectionId);
      }
    };

    ws.onerror = () => {
      callbacks.onError(new Error('Quick action WebSocket connection error'));
      this.activeConnections.delete(connectionId);
    };

    ws.onclose = () => {
      this.activeConnections.delete(connectionId);
    };

    this.activeConnections.set(connectionId, ws);
  }

  /**
   * Dispose and clean up
   */
  dispose(): void {
    for (const ws of this.activeConnections.values()) {
      ws.close();
    }
    this.activeConnections.clear();
  }
}
